README.md
# Description

The cmd-wheat-detection Python package includes code to train and evaluate a basic model for the [Kaggle Global Wheat Detection](https://www.kaggle.com/c/global-wheat-detection/overview) problem. This deep learning model uses [pytorch](https://pytorch.org/) with [pytorch lightning](https://www.pytorchlightning.ai/) to run [torchvision](https://pytorch.org/vision/stable/index.html) object detection models.

You can see how I use this code in a Kaggle notebook [here](https://www.kaggle.com/sheromon/wheat-faster-r-cnn-with-pytorch-lightning/notebook).

# Usage

Set up virtual environment
```
python3 -m venv venv
source venv/bin/activate
pip install --upgrade pip
pip install -e .
```

Activate virtual environment
```
source venv/bin/activate
```

## Command-line scripts
Run training as a script using the default configuration
```
train  # using python entry point
```
or
```
python wheat/scripts/train.py
```

As documented in the [PyTorch Lightning docs](https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html#trainer-flags), PyTorch Lightning CLI flags can be used to configure the training run. For example, to set the number of training epochs for to 10, add `--max_epochs=10` to the train command.

Run evaluation on the validation set with weights from a saved checkpoint
```
evaluate <path-to-checkpoint> # using python entry point
```
or
```
python wheat/scripts/evaluate.py <path-to-checkpoint>
```

Run inference on the test set and save predictions in the Kaggle submission format
```
predict <path-to-checkpoint> # using python entry point
```
or
```
python wheat/scripts/predict.py <path-to-checkpoint>
```

## Jupyter notebook
To install Jupyter notebook support, with the virtual environment active, run `pip install -e .[notebook]`, then start Jupyter.

### Provided notebooks
  * Visualize dataset examples: [notebooks/visualize_dataset.ipynb](notebooks/visualize_dataset.ipynb)
  * Run training: [notebooks/train.ipynb](notebooks/train.ipynb)
  * Run evaluation and analyze results: [notebooks/evaluate.ipynb](notebooks/evaluate.ipynb)

## Configuration

Configuration is managed using [configobj](https://configobj.readthedocs.io/en/latest/configobj.html). The full set of configurable parameters with type information is available in [wheat/config/configspec.ini]([wheat/config/configspec.ini]). Overrides to default values can be specified in a custom `config.ini` file.

### Example custom config file
Indentation is not required but is included because I find it helps me see which parameters are in which section or sub-section.
```
['train']
    batch_size = 2
    [['optimizer']]
        initial_lr = 0.0001
```

Run training with a custom configuration
```
train --config-path my_config.ini
```

# TODO

Honestly, I'm not sure when I'll have time to do any of these, but it's nice to have goals, right?

* Get rid of "numpy array is not writeable" error
* Replace config with PyTorch Lightning hyperparameters
* Maybe replace numpy seed with PyTorch Lightning seed feature
* Add stochastic weight averaging
* Add tests?
* Add test-time augmentation?


CHANGELOG.md
# 0.2.0

* Updated Python packaging to include configspec.ini and config.ini as data files in the package and updated code to use installed .ini files
* Fixed bug in format of submission.csv (was missing header row)

# 0.1.0

* Created a PyTorch Lightning data module to work with the Kaggle Global Wheat Detection dataset
* Set up a PyTorch Lightning model to wrap the torchvision Faster R-CNN object detector
* Set up to allow creating a pip installable Python package
* Created Jupyter notebooks for dataset visualization, training, and evaluation
* Created Python entry points for training, evaluation, and inference
* Added cyclic learning rate scheduler


wheat/dataset.py
"""PyTorch Dataset class for loading wheat images and annotations"""
from pathlib import Path

import numpy as np
from PIL import Image
import torch
from torch.utils.data import Dataset
import torchvision.transforms as T


class WheatDataset(Dataset):
    """PyTorch Dataset class for loading wheat images and annotations"""

    def __init__(self, image_dir: Path, image_ids: list, anno_dict: dict, transform=None):
        """Store image and annotation data and corresponding transforms.

        :param image_dir: path to directory containing image files
        :param image_ids: list of image IDs (image file name stems)
        :param anno_dict: dictionary with image ID as key and list of bounding
            boxes [xmin, ymin, width, height] as values
        :param transform: albumentations PyTorch transform to be applied to
            images and bounding boxes
        """
        self.image_dir = image_dir
        self.image_ids = image_ids
        self.anno_dict = {key: val for key, val in anno_dict.items() if key in image_ids}
        # TODO enforce that transform is an albumentations transform
        self.transform = transform

        if not all((image_id in self.anno_dict for image_id in self.image_ids)):
            raise RuntimeError('Not all image IDs were found in anno_dict.')

    def __len__(self):
        return len(self.image_ids)

    def __getitem__(self, ind):
        """Return one dataset sample from given index."""
        if torch.is_tensor(ind):
            ind = ind.tolist()
        image_id = self.image_ids[ind]
        image_path = self.image_dir/(image_id + '.jpg')
        image = np.asarray(Image.open(image_path))

        target_boxes = []
        for bbox in self.anno_dict[image_id]:
            # input bbox is [xmin, ymin, width, height]
            left = bbox[0]
            top = bbox[1]
            right = bbox[0] + bbox[2]
            bottom = bbox[1] + bbox[3]
            # output bbox is [xmin, ymin, xmax, ymax]
            target_boxes.append([left, top, right, bottom])

        # if no annos then boxes tensor will be wrong shape, so fix the shape
        boxes_tensor = np.array(target_boxes, dtype=np.float32) \
            if target_boxes else np.zeros((0, 4), dtype=np.float32)
        n_boxes = boxes_tensor.shape[0]
        target = {
            'boxes': boxes_tensor,
            # there's only one class, so class labels are all ones
            'labels': np.ones(n_boxes, dtype=np.int64),
        }

        if self.transform is not None:
            # apply albumentations transformations
            transformed = self.transform(
                image=image,
                bboxes=target['boxes'],
                labels=target['labels'],
            )
            # albumentations returns the transformed image in uint8 format, but
            # torchvision wants it as a float, so convert it
            image_int_to_float_transform = T.ConvertImageDtype(torch.float)
            image = image_int_to_float_transform(transformed['image'])
            # the albumentations bounding box transforms only convert the images
            # to tensors, so we have to convert the boxes and labels to tensors
            target = {
                'boxes': torch.tensor(np.array(transformed['bboxes'], dtype=np.float32)) \
                    if transformed['bboxes'] else torch.zeros((0, 4), dtype=torch.float32),
                'labels': torch.tensor(np.array(transformed['labels'], dtype=np.int64)),
            }

        return image, target


wheat/data_module.py
"""PyTorch Lightning DataModule for working with the Kaggle wheat dataset"""
import ast
import os
from pathlib import Path
from typing import Optional

import albumentations as A
import albumentations.pytorch
import numpy as np
import pandas as pd
from pytorch_lightning import LightningDataModule
from torch.utils.data import DataLoader

from wheat.dataset import WheatDataset


class WheatDataModule(LightningDataModule):
    """PyTorch Lightning DataModule for working with the Kaggle wheat dataset"""
    def __init__(self, config):
        super().__init__()
        self.config = config
        self.train_dataset = None
        self.val_dataset = None

    def prepare_data(self):
        pass

    def get_transforms(self):
        """Construct and return transforms for train and val/test/predict stages.

        Training trainsforms are composed albumentations transforms with
        probabilities controlled in the config. Val/test/predict transform is
        just converting the images to torch tensors.

        :return: tuple (train_transform, val_transform)
        """
        train_trans_config = self.config['train']['transforms']
        bbox_params = A.BboxParams(
            format='pascal_voc', min_visibility=0.5, label_fields=['labels'])
        train_transform = A.Compose([
            A.HorizontalFlip(p=train_trans_config['horizontal_flip_prob']),
            A.VerticalFlip(p=train_trans_config['vertical_flip_prob']),
            A.RandomRotate90(p=train_trans_config['random_rotate90_prob']),
            A.ColorJitter(p=train_trans_config['color_jitter_prob']),
            A.Blur(p=train_trans_config['blur_prob']),
            albumentations.pytorch.ToTensorV2(),
        ], bbox_params=bbox_params)
        val_transform = A.Compose([
            albumentations.pytorch.ToTensorV2(),
        ], bbox_params=bbox_params)
        return train_transform, val_transform

    def setup(self, stage: Optional[str] = None):
        """Load annotation data, create train/val split, and create datasets.

        :param stage: one of 'fit', 'validate', 'test', or 'predict'
        """
        data_dir = Path(self.config['data_dir'])
        train_or_test = 'train' if stage in ('fit', 'validate') else 'test'
        image_dir = data_dir/train_or_test
        # must get unique image ids from files on disk, not from train.csv,
        # because images without bounding boxes are not in train.csv
        unique_image_ids = sorted([image_path.stem for image_path in image_dir.glob('*.jpg')])
        anno_dict = {image_id: [] for image_id in unique_image_ids}

        train_transform, val_transform = self.get_transforms()
        if stage in ['fit', 'validate']:
            train_csv = data_dir/'train.csv'
            df = pd.read_csv(train_csv)
            for _, row in df.iterrows():
                bbox = np.array(ast.literal_eval(row.bbox))
                anno_dict[row.image_id].append(bbox)

            rng = np.random.default_rng(seed=self.config['numpy_seed'])
            rng.shuffle(unique_image_ids)
            train_fraction = 0.8
            train_samples = int(train_fraction * len(unique_image_ids))
            train_ids = unique_image_ids[:train_samples]
            val_ids = unique_image_ids[train_samples:]
            # if we want to run overfitting, use the val set as the train set
            if self.config['train']['overfit']:
                train_ids = val_ids

            if stage == 'fit':
                self.train_dataset = WheatDataset(
                    image_dir, train_ids, anno_dict, transform=train_transform)
            self.val_dataset = WheatDataset(
                image_dir, val_ids, anno_dict, transform=val_transform)

        if stage == 'predict':
            self.test_dataset = WheatDataset(
                image_dir, unique_image_ids, anno_dict, transform=val_transform)

    def train_dataloader(self):
        return DataLoader(
            self.train_dataset,
            batch_size=self.config['train']['batch_size'],
            shuffle=True,
            num_workers=os.cpu_count(),
            collate_fn=collate_fn,
        )

    def val_dataloader(self):
        return DataLoader(
            self.val_dataset,
            batch_size=self.config['eval']['batch_size'],
            num_workers=os.cpu_count(),
            collate_fn=collate_fn,
        )

    def predict_dataloader(self):
        return DataLoader(
            self.test_dataset,
            batch_size=1,
            num_workers=os.cpu_count(),
            collate_fn=collate_fn,
        )

def collate_fn(batch):
    """From https://github.com/pytorch/vision/blob/main/references/detection/utils.py"""
    return tuple(zip(*batch))


wheat/scripts/train.py
"""Training script"""
import argparse

import pytorch_lightning as pl

from wheat.config import load_config
from wheat.data_module import WheatDataModule
from wheat.model import WheatModel


def train(config, args_dict):
    wheat_data_module = WheatDataModule(config)
    model = WheatModel(config)

    checkpointer = pl.callbacks.ModelCheckpoint(monitor='ap75', mode='max')
    lr_monitor = pl.callbacks.LearningRateMonitor(logging_interval='epoch', log_momentum=True)
    trainer = pl.Trainer(**args_dict, callbacks=[checkpointer, lr_monitor])
    trainer.fit(model, wheat_data_module)
    trainer.validate(model, wheat_data_module)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--config-path', type=str, default=None,
                        help='Path to config file')
    parser = pl.Trainer.add_argparse_args(parser)
    args = parser.parse_args()

    args_dict = vars(args)
    config = load_config(args_dict.pop('config_path'))
    train(config, args_dict)


if __name__ == '__main__':
    main()


wheat/visualization.py
"""A couple of functions for visualizing the dataset in Jupyter notebooks"""
import matplotlib.pyplot as plt
import numpy as np
import torch
import torchvision.transforms as T
import torchvision.transforms.functional as F


# this conversion is needed because albumentations transforms return
# images in uint8, but pytorch expects them to be floats in [0, 1]
image_float_to_int_transform = T.ConvertImageDtype(torch.uint8)

def show(imgs):
    """Display a single or a list of torch tensor images.

    from https://pytorch.org/vision/stable/auto_examples/plot_visualization_utils.html
    """
    if not isinstance(imgs, list):
        imgs = [imgs]
    _, axs = plt.subplots(ncols=len(imgs), squeeze=False)
    for i, img in enumerate(imgs):
        img = img.detach()
        img = F.to_pil_image(img)
        axs[0, i].imshow(np.asarray(img))
        axs[0, i].set(xticklabels=[], yticklabels=[], xticks=[], yticks=[])


wheat/config/__init__.py
"""Config loading and validation"""
from pathlib import Path

import configobj
from validate import Validator

# TODO: use ResourceManager API as recommended in
# https://setuptools.pypa.io/en/latest/userguide/datafiles.html
CONFIG_DIR = Path(__file__).parent


def load_config(config_path=None, configspec_path=None, do_validation=True):
    """Load config.

    :param str config_path: optional path to configuration file with
        model-specific parameters; if not provided, load default config.ini
    :param str configspec_path: optional path to model configspec file
    :param bool do_validation: optional; if True (default), run validation

    :return: ConfigObj object, basically a dictionary of parameters
    """
    if config_path is None:
        config_path = CONFIG_DIR/'config.ini'
    else:
        config_path = Path(config_path)
    if not config_path.is_file():
        raise FileNotFoundError(f"No file '{config_path}'.")
    # Assumes that the configspec.ini for the given config file is in the same directory
    if not configspec_path:
        configspec_path = config_path.parent/'configspec.ini'

    spec = configobj.ConfigObj(str(configspec_path), list_values=False)
    config = configobj.ConfigObj(str(config_path), configspec=spec)

    # Read config file and assign appropriate types to values
    validator = Validator()

    result = config.validate(validator, preserve_errors=True)
    if do_validation and result is not True:
        for key, value in result.items():
            if value is not True:
                raise RuntimeError(f'Config parameter "{key}": {value}')

    return config


wheat/scripts/predict.py
"""Prediction script"""
import argparse

import pytorch_lightning as pl

from wheat.config import load_config
from wheat.data_module import WheatDataModule
from wheat.dataset import WheatDataset
from wheat.model import WheatModel


def predict(config, args_dict: dict, ckpt_path: str):
    """Run inference on the test set and save predictions.

    :param config: configobj mapping of config paramters to values
    :param args_dict: dict of options for initializing PyTorch Lightning Trainer
    :param ckpt_path: path to checkpoint to load for inference
    """
    wheat_data_module = WheatDataModule(config)
    model = WheatModel(config)
    trainer = pl.Trainer(**args_dict)
    results = trainer.predict(model, wheat_data_module, ckpt_path=ckpt_path)
    score_threshold = config['predict']['score_threshold']
    save_submission(results, wheat_data_module.test_dataset, score_threshold)


def save_submission(results, test_dataset: WheatDataset, score_threshold: float,
                    save_path='submission.csv'):
    """Save inference outputs in format specified by Kaggle competition.

    :param results: outputs from Trainer.predict; list of list of dicts with keys
        'scores', 'boxes', and 'labels'
    :param score_threshold: predictions with scores below this value will be discarded
    """
    image_ids = test_dataset.image_ids
    header = 'image_id,PredictionString\n'
    with open(save_path, 'w', encoding='utf-8') as file_obj:
        file_obj.write(header)
        for image_id, predictions in zip(image_ids, results):
            keep_inds = predictions[0]['scores'] > score_threshold
            keep_scores = (predictions[0]['scores'][keep_inds]).cpu().numpy()
            keep_boxes = predictions[0]['boxes'][keep_inds, :].cpu().numpy()
            # convert last two columns to width and height instead of xmax, ymax
            keep_boxes[:, 2:] -= keep_boxes[:, :2]
            line = f'{image_id},'
            for ind, score in enumerate(keep_scores):
                line += f' {score} '
                line += ' '.join([str(round(val)) for val in keep_boxes[ind]])
            line += '\n'
            file_obj.write(line)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('ckpt_path', type=str,
                        help='Path to model checkpoint file')
    parser.add_argument('--config-path', type=str, default=None,
                        help='Path to config file')
    parser = pl.Trainer.add_argparse_args(parser)
    args = parser.parse_args()

    args_dict = vars(args)
    config = load_config(args_dict.pop('config_path'))
    ckpt_path = args_dict.pop('ckpt_path')
    predict(config, args_dict, ckpt_path)


if __name__ == '__main__':
    main()


wheat/scripts/evaluate.py
"""Evaluation script"""
import argparse

import pytorch_lightning as pl

from wheat.config import load_config
from wheat.data_module import WheatDataModule
from wheat.model import WheatModel


def evaluate(config, args_dict, ckpt_path):
    """Run evaluation on the val set.

    :param config: configobj mapping of config paramters to values
    :param args_dict: dict of options for initializing PyTorch Lightning Trainer
    :param ckpt_path: path to checkpoint to load for inference
    """
    wheat_data_module = WheatDataModule(config)
    model = WheatModel(config)
    trainer = pl.Trainer(**args_dict)
    trainer.validate(model, wheat_data_module, ckpt_path=ckpt_path)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('ckpt_path', type=str,
                        help='Path to model checkpoint file')
    parser.add_argument('--config-path', type=str, default=None,
                        help='Path to config file')
    parser = pl.Trainer.add_argparse_args(parser)
    args = parser.parse_args()

    args_dict = vars(args)
    config = load_config(args_dict.pop('config_path'))
    ckpt_path = args_dict.pop('ckpt_path')
    evaluate(config, args_dict, ckpt_path)


if __name__ == '__main__':
    main()


wheat/model.py
"""PyTorch Lightning Module with torchvision object detection model"""
import os
from pathlib import Path

import brambox
import numpy as np
import pandas as pd
import pytorch_lightning as pl
import torch
from torchvision.models.detection import faster_rcnn


class WheatModel(pl.LightningModule):
    """PyTorch Lightning Module with torchvision object detection model"""

    def __init__(self, config):
        super().__init__()
        self._loss_names = (
            'loss_classifier', 'loss_box_reg',
            'loss_objectness', 'loss_rpn_box_reg',
        )
        self.config = config
        self.learning_rate = self.config['train']['optimizer']['initial_lr']
        # num_classes includes background, so wheat and background are two classes
        self.model = faster_rcnn.fasterrcnn_resnet50_fpn(num_classes=2)

    def forward(self, example):
        image, targets = example
        # pytorch lightning wants the loss calculate to be in the training step,
        # not in the forward method, but the torchvision models don't cooperate,
        # so here we are returning losses in forward. note that in eval mode, we
        # will be returning a list of predictions instead of losses.
        return self.model(image, targets)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(
            self.parameters(),
            lr=self.learning_rate,
            momentum=0.8,
        )
        lr_scheduler = torch.optim.lr_scheduler.CyclicLR(
            optimizer,
            base_lr=self.config['train']['optimizer']['initial_lr'],
            max_lr=self.config['train']['optimizer']['max_lr'],
        )
        lr_scheduler_config = {
            'scheduler': lr_scheduler,
            'interval': 'step',
            'frequency': 1,
        }
        return {'optimizer': optimizer, 'lr_scheduler': lr_scheduler_config}

    def training_step(self, batch, batch_idx):  # pylint: disable=unused-argument
        """Run the model in training mode.

        :return: scalar tensor with total loss value
        """
        # note from pytorch docs: Prior to PyTorch 1.1.0, the learning rate
        # scheduler was expected to be called before the optimizer’s update;
        # #1.1.0 changed this behavior in a BC-breaking way.
        loss_dict = self(batch)
        total_loss = sum([loss_dict[key] for key in self._loss_names])
        loss_dict['total_loss'] = total_loss
        self.log_dict(loss_dict, on_step=True, batch_size=len(batch[1]))
        return total_loss

    def validation_step(self, batch, batch_idx):  # pylint: disable=unused-argument
        """Run the model in eval mode.

        :return: tuple (labels, predictions) where labels and predictions are
            lists with length equal to the batch size; each element of labels is
            a dict with keys 'boxes' and 'labels', and eachelement of predictions
            is a dict with keys 'boxes', 'labels', and 'scores'
        """
        # torchvision generalized r-cnn models return different outputs when
        # running in train mode versus eval mode. I wish the forward method just
        # returned raw predictions instead of the full process of calculating
        # loss or creating predicted bounding boxes, but hey, it's free code,
        # and beggars can't be choosers.
        predictions = self(batch)
        # batch[0] is a list of input images, and we don't need those for eval
        return batch[1], predictions

    def validation_epoch_end(self, outputs):
        """Calculate average precision (AP) and log it."""
        det_df, anno_df = self.compile_eval_dataframes(outputs)
        # if these EVs are set, write the dataframes to disk
        # this lets us do offline analysis later on
        save_dir = os.environ.get('CMD_WHEAT_OUTPUT_DIR')
        print('save_dir:', save_dir)
        if save_dir:
            det_save_path = Path(save_dir)/'det.csv'
            anno_save_path = Path(save_dir)/'anno.csv'
            det_df.to_csv(det_save_path)
            anno_df.to_csv(anno_save_path)
        # kaggle global wheat detection evaluation metric is AP calculated from
        # 0.5 to 0.75 IOU in 0.05 increments. I don't want to calculate AP at
        # every value, so just calculate it at 0.5 and 0.75.
        for iou_threshold in [0.5, 0.75]:
            df_pr = brambox.stat.pr(det_df, anno_df, threshold=iou_threshold)
            ap = brambox.stat.ap(df_pr)
            metric_name = 'ap' + str(round(100*iou_threshold))
            self.log(metric_name, ap, on_epoch=True)

    def compile_eval_dataframes(self, outputs):
        """Reformat eval outputs into detection and annotation dataframes.

        This output format is consistent with the required input format for the
        brambox package.

        :return: tuple of pandas dataframes (det_df, anno_df), where each dataframe
            has columns that define the ground truth or predicted bounding box
            'x_top_left', 'y_top_left', 'width', 'height'; an image ID 'image';
            a class label 'class_label', and for detections, a score 'confidence'
        """
        anno_df_list = []
        det_df_list = []
        for ibatch, val_outputs_batch in enumerate(outputs):
            for isample, (anno, pred) in enumerate(zip(*val_outputs_batch)):
                ind = ibatch * len(val_outputs_batch[0]) + isample
                anno_data = {
                    'image': ind * np.ones(len(anno['labels']), dtype=np.int64),
                    'x_top_left': anno['boxes'][:, 0].cpu().numpy().astype(np.float64),
                    'y_top_left': anno['boxes'][:, 1].cpu().numpy().astype(np.float64),
                    'width': (anno['boxes'][:, 2] - anno['boxes'][:, 0]).cpu() \
                        .numpy().astype(np.float64),
                    'height': (anno['boxes'][:, 3] - anno['boxes'][:, 1]).cpu() \
                        .numpy().astype(np.float64),
                    'class_index': anno['labels'].cpu().numpy().astype(np.int64),
                }
                anno_df = pd.DataFrame(data=anno_data)
                anno_df['class_label'] = anno_df['class_index']
                anno_df['ignore'] = False
                anno_df_list.append(anno_df)

                pred_data = {
                    'image': ind * np.ones(len(pred['labels']), dtype=np.int64),
                    'x_top_left': pred['boxes'][:, 0].cpu().numpy().astype(np.float64),
                    'y_top_left': pred['boxes'][:, 1].cpu().numpy().astype(np.float64),
                    'width': (pred['boxes'][:, 2] - pred['boxes'][:, 0]).cpu() \
                        .numpy().astype(np.float64),
                    'height': (pred['boxes'][:, 3] - pred['boxes'][:, 1]).cpu() \
                        .numpy().astype(np.float64),
                    'class_index': pred['labels'].cpu().numpy().astype(np.int64),
                    'confidence': pred['scores'].cpu().numpy().astype(np.float64),
                }
                det_df = pd.DataFrame(data=pred_data)
                det_df['class_label'] = det_df['class_index']
                det_df_list.append(det_df)
        return pd.concat(det_df_list), pd.concat(anno_df_list)
